File: coding/src/com/algos/core/li12_trie/revised/L1858_LongestWordWithAllPrefixes.java
package  com.algos.core.li12_trie.revised;

import com.algos.core.models.TrieNode;

public class L1858_LongestWordWithAllPrefixes {
    public static void main(String[] args) {
        L1858_LongestWordWithAllPrefixes obj = new L1858_LongestWordWithAllPrefixes();

        String[] words = new String[]{"a", "banana", "app", "appl", "ap", "apply", "apple"};
        System.out.println(obj.longestWord(words));
    }

    // Problem: https://leetcode.com/problems/longest-word-with-all-prefixes/
    // https://leetcode.ca/2021-07-14-1858-Longest-Word-With-All-Prefixes/
    public String longestWord(String[] words) {
        TrieNode root = new TrieNode();
        for(String word: words)
            insert(root, word);

        String longest = "";
        for(String word: words){
            if(checkIfPrefixesExists(root, word)){
                if(word.length() > longest.length()){
                    longest = word;
                }else if(word.length() == longest.length() && word.compareTo(longest) < 0){
                    longest = word;
                }
            }
        }
        return longest;
    }

    public void insert(TrieNode root, String word) {
		TrieNode curr = root;
        for(Character ch : word.toCharArray()){
            if(!curr.map.containsKey(ch)){
                TrieNode newNode = new TrieNode();
                curr.map.put(ch, newNode);
            }
            curr = curr.map.get(ch);
        }
        curr.endOfWord = true;	// after new node is created current will be pointing to new node after for loop
	}

    private boolean checkIfPrefixesExists(TrieNode root, String word) {
        TrieNode curr = root;
        boolean flag = true;
        for(Character ch : word.toCharArray()){
            if(!curr.map.containsKey(ch)){
                return false;
            }
            curr = curr.map.get(ch);
            flag = flag&curr.endOfWord;
        }
        return flag;
    }
}

File: dsa/patterns/12_tries/python/l1858_longest_word_with_all_prefixes.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False


"""
Problem: https://leetcode.com/problems/longest-word-with-all-prefixes/
https://leetcode.ca/2021-07-14-1858-Longest-Word-With-All-Prefixes/

Idea:
Insert all words into a Trie.
For each word, check if *all its prefixes* exist in the trie as complete words.
Choose the longest such word.
If multiple words have the same length, choose lexicographically smaller one.
"""


def longest_word(words):
    root = TrieNode()

    # Insert all words into trie
    for word in words:
        insert(root, word)

    longest = ""
    for word in words:
        if prefixes_exist(root, word):
            if len(word) > len(longest):
                longest = word
            elif len(word) == len(longest) and word < longest:
                longest = word

    return longest


def insert(root, word):
    curr = root
    for ch in word:
        if ch not in curr.children:
            curr.children[ch] = TrieNode()
        curr = curr.children[ch]
    curr.end_of_word = True


def prefixes_exist(root, word):
    curr = root
    for ch in word:
        if ch not in curr.children:
            return False
        curr = curr.children[ch]
        if not curr.end_of_word:
            return False
    return True


if __name__ == "__main__":
    words = ["a", "banana", "app", "appl", "ap", "apply", "apple"]
    print(longest_word(words))   # apple
